### ISSUE TITLE
CH vs astar routing leads to substantially different results

### ISSUE THREAD

**Issue by ambuehll** [eclipse-sumo/sumo#9803]
This is a follow-up on what was discussed this morning via chat. 

I came across _substantial_ discrepancies in network traffic performance depending on the router chosen.
I have a larger network (passenger vType only), where demand and all other parameters are identical except for the routing algorithm. 
Average flow in the network is substantially higher with astar (100 veh/h-lane) and the network does not gridlock as for CH. 

![image](https://user-images.githubusercontent.com/48930860/146213887-622fd76c-23dd-453a-a0c0-49a1f706ca27.png) 
![image](https://user-images.githubusercontent.com/48930860/146214334-f13a750f-abc0-4e82-8bff-3e9487c5b4ff.png)

It seems as if reroutes are not triggered very often in CH compared to astar. For my scenario: CH: only 11 vehicles out of 123'000 vs astar  9'000 of 130'000 vehicles get rerouted.

I checked some vehicle routes. For example: "D19184" only gets 1 CH route, whereas it gets rerouted once in the astar case. 
The cost of its chosen CH route in the penultimate time interval (at 32220s) is 26522 for the whole route, but if it were to reroute to the astar alternative, it would reduce to 8644.
Important to note: I found a discrepancy between the "future" routing cost evaluated at 32220s (cost occurred so far + cost of any planned edge at 32220s) and the one recorded in the route file (5336) (does this value get updated dynamically if reroute prob is 1?)
Here are the [files](https://drive.google.com/drive/folders/16LcGLIK5wDcoBN5ozv2HwcRcfGzLIu6I?usp=sharing). 
Let me know if there is more info necessary. 

**SUMO-version:**v1_11_0+0421-96d1c786f0c

**operating system:**Windows

**Comment by ambuehll** [comment]
Hi Michael, 
Sorry to bring this up again. I was wondering whether this was still on the list for the next release. Would very much appreciate that. I did look into this again, and this issue is unfortunate as CH scales much better with a higher number of vehicles. 
Many thanks,
Lukas

**Comment by namdre** [comment]
There is something obviously fishy with the cost computation: CH records rerouting with negative savings!

```
attributeStats.py -e route -a savings astar.routes_ZRH.xml 
route savings: count 92317, min 0.00, max 134.25, mean 0.35, Q1 0.00, median 0.00, Q3 0.00,  stdDev  3.12

attributeStats.py -e route -a savings CH.routes_ZRH.xml 
route savings: count 85890, min -177.84, max 13.17, mean -0.00, Q1 0.00, median 0.00, Q3 0.00,  stdDev  0.86
```

**Comment by behrisch** [comment]
This is not completely unexpected since the savings are calculated with respect to the current state while the routing is done on a maybe slightly outdated hierarchy

**Comment by namdre** [comment]
as far as I can tell, the hierarchy update should be in sync with the routing device adaptation-interval. At least the frequency of hierarchy rebuilding matches the adaptation interval when testing this.

**Comment by behrisch** [comment]
In theory Jakob is right, however I noticed today that it does not make a difference whether I enable or disable the hierarchy rebuilding. So there is obviously something wrong with the hierarchy builder here.

**Comment by behrisch** [comment]
CHRouter did not use the correct travel times so it did not find new routes. This has been fixed now. It is still unclear why it did find different routes anyway and there are still negative savings so it is not over yet (but better than before ;-)).

**Comment by behrisch** [comment]
Some further remarks to the scenario:

1. We found in our tests that synchronized rerouting leads to more artifacts because sometimes all vehicles approaching a jam switch the route at once inducing a new jam.
2. It may be a waste of time to update the edge weights more often than you do the rerouting (especially if you do synchronized rerouting). If you have different data here, I would be interested

**Comment by ambuehll** [comment]
Many thanks for looking into this. I'll come back with performance values with this semi-fix. 

Ad 1. I see, I was hoping that the additional weight.random-factor would help with that. I'll double check what the performance effects are when disabling synchronisation.

Ad 2. Never intended to have different values, this was due to some tests I did back then where I changed the numbers frequently but forgot to change one back. Completely understand that this is not a very smart thing to do..

Again, many thanks as always!

**Comment by behrisch** [comment]
I am not sure that using weight.random-factor together with CH is a good idea. The randomness is not consistent over several calls of getEffort which means you get different weights for the same edge.

**Comment by ambuehll** [comment]
to clarify, are you referring to https://github.com/eclipse/sumo/issues/6808#issuecomment-599895090 or something else?

**Comment by behrisch** [comment]
Yes, at least in spirit it is the same issue. As long as weight.random-factor is != 1, every time you retrieve the cost for an edge a different result will appear, even if it is the same vehicle and the same time step. This is not so critical for astar and dijkstra since they usually query each edge cost only once. It is probably worse for CH since it uses multiple shortest path trees on hierarchy building.

**Comment by namdre** [comment]
I think a better solution might be to use the non-randomized weights when building the hierarchy and then only randomizing at query time.

Anyway, the randomization should fully explain the negative savings so I'd consider this issue closed.

It probably makes sense to also compute the savings with the non-randomized weights (this keeps the negative savings but makes the value more meaningful). Again, a use case of using both randomized and non-randomized weights at the same time.

**Comment by behrisch** [comment]
Unfortunately it does not explain it because in the tests randomization was disabled.

**Comment by namdre** [comment]
after the fixes from @behrisch the results are still different enough to be suspicious. Simulating until 8:00 gives:

| algo  | astar  | CH  |
|---|---|---|
| Inserted                         | 87631    |87528     |
| Running                          | 7987     |8808      |
| Waiting                          | 42       |71        |
| Teleports                        | 349      |390       |
| count                            | 79644    |78720     |
| RouteLength                      | 5860.72  |5851.58   |
| Speed                            | 10.89    |10.70     |
| Duration                         | 487.11   |504.75    |
| WaitingTime                      | 7.71     |9.56      |
| TimeLoss                         | 101.25   |119.12    |
| DepartDelay                      | 0.90     |1.05      |

**Comment by ambuehll** [comment]
I found similar discrepancies, but I didn't qualify them as "different enough". To me, they seem in the range of https://github.com/eclipse/sumo/issues/6319. But I'll run a deeper investigation once I find some time. Thanks!
